File: world/examples/cartpole_vpg.py
"""
CartPole-v1 VPG (Vanilla Policy Gradient) with ai4/nn
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
REINFORCE algorithm implementation using ai4.nn framework

Run:
    PYTHONPATH=. python world/examples/cartpole_vpg_ai4nn.py
"""

import itertools
import logging
import numpy as np

import world
from nn import Tensor, Module, Linear, Sequential, ReLU, Adam

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s [%(levelname)s] %(message)s',
    stream=__import__('sys').stdout,
    datefmt='%H:%M:%S'
)

env = world.make('CartPole-v1')


class PolicyNet(Module):
    def __init__(self, obs_dim: int, action_n: int):
        super().__init__()
        self.net = Sequential(
            Linear(obs_dim, 128),
            ReLU(),
            Linear(128, 64),
            ReLU(),
            Linear(64, action_n),
        )

    def __call__(self, x):
        return self.forward(x)

    def forward(self, x):
        logits = self.net(x)
        return logits.softmax(axis=1)


class VPGAgent:
    def __init__(self, env):
        self.action_n = env.action_space.n
        self.gamma = 0.99
        self.obs_dim = env.observation_space.shape[0]
        
        self.policy_net = PolicyNet(self.obs_dim, self.action_n)
        self.optimizer = Adam(self.policy_net.parameters(), lr=0.005)

    def reset(self, mode=None):
        self.mode = mode
        if self.mode == 'train':
            self.trajectory = []

    def step(self, observation, reward, terminated):
        state = Tensor(observation.reshape(1, -1), requires_grad=False)
        probs = self.policy_net(state)
        
        # Sample action from categorical distribution
        probs_np = probs.data.flatten()
        action = np.random.choice(self.action_n, p=probs_np)
        
        if self.mode == 'train':
            self.trajectory += [observation, reward, terminated, action]
        return action

    def close(self):
        if self.mode == 'train':
            self.learn()

    def learn(self):
        # Extract trajectory
        states = [self.trajectory[i] for i in range(0, len(self.trajectory), 4)]
        rewards = [self.trajectory[i] for i in range(1, len(self.trajectory), 4)]
        actions = [self.trajectory[i] for i in range(3, len(self.trajectory), 4)]
        
        # Compute discounted returns
        T = len(rewards)
        discounted_returns = np.zeros(T, dtype=np.float32)
        running = 0.0
        for t in reversed(range(T)):
            running = rewards[t] + self.gamma * running
            discounted_returns[t] = running
        
        discounted_returns = discounted_returns / (discounted_returns.std() + 1e-8)
        
        # Compute loss and gradients using numerical gradient
        params = self.policy_net.parameters()
        eps = 1e-5
        
        for p in params:
            grad = np.zeros_like(p.data)
            original = p.data.copy()
            
            for i in range(p.data.size):
                p.data.flat[i] += eps
                
                loss_plus = self._compute_loss(states, actions, discounted_returns)
                
                p.data.flat[i] = original.flat[i] - eps
                loss_minus = self._compute_loss(states, actions, discounted_returns)
                
                grad.flat[i] = (loss_plus - loss_minus) / (2 * eps)
                p.data.flat[i] = original.flat[i]
            
            p.grad = grad
        
        self.optimizer.step()

    def _compute_loss(self, states, actions, discounted_returns):
        T = min(len(states), 100)
        loss = 0.0
        states_arr = np.array(states[:T], dtype=np.float32)
        
        for t in range(T):
            state_t = Tensor(states_arr[t:t+1], requires_grad=False)
            probs_t = self.policy_net(state_t)
            log_prob_t = np.log(probs_t.data[0, actions[t]] + 1e-8)
            loss -= discounted_returns[t] * log_prob_t
        
        return loss / T


agent = VPGAgent(env)


def play_episode(env, agent, seed=None, mode=None, render=False):
    observation, info = env.reset(seed=seed)
    reward, terminated, truncated = 0., False, False
    agent.reset(mode=mode)
    episode_reward, elapsed_steps = 0., 0
    
    while True:
        action = agent.step(observation, reward, terminated)
        if render:
            env.render(mode="ansi")
        if terminated or truncated:
            break
        observation, reward, terminated, truncated, info = env.step(action)
        episode_reward += reward
        elapsed_steps += 1
    agent.close()
    return episode_reward, elapsed_steps


# Training
logging.info('==== train ====')
episode_rewards = []
for episode in itertools.count():
    episode_reward, elapsed_steps = play_episode(env, agent, seed=episode, mode='train')
    episode_rewards.append(episode_reward)
    if episode % 50 == 0:
        logging.info('train episode %d: reward = %.2f, steps = %d',
                     episode, episode_reward, elapsed_steps)
    if np.mean(episode_rewards[-20:]) > 199:
        break

logging.info(f"Training complete! Total episodes: {len(episode_rewards)}")
logging.info(f"Average reward (last 20): {np.mean(episode_rewards[-20:]):.2f}")

# Test
logging.info('==== test ====')
test_rewards = []
for episode in range(100):
    episode_reward, _ = play_episode(env, agent)
    test_rewards.append(episode_reward)

logging.info('average episode reward = %.2f ± %.2f',
             np.mean(test_rewards), np.std(test_rewards))

File: world/examples/cartpole_vpg.ts
/**
 * examples/cartpole_vpg.ts
 * VPG (Vanilla Policy Gradient / REINFORCE) for CartPole-v1
 * using ai4 nn framework with numerical gradients.
 *
 * The policy network is a 3-layer MLP: 4 → 128 → 64 → 2 (softmax).
 * Training uses REINFORCE with discounted returns and Adam optimizer.
 *
 * Usage:
 *   npx tsx world/examples/cartpole_vpg.ts
 */
import { CartPoleEnv } from '../envs/cartpole';
import { Tensor, Module, Linear, ReLU, Sequential, Adam } from '../../nn/index';

class PolicyNet extends Module {
  net: Sequential;

  constructor(obsDim: number, actionN: number) {
    super();
    this.net = new Sequential([
      new Linear(obsDim, 128, true),
      new ReLU(),
      new Linear(128, 64, true),
      new ReLU(),
      new Linear(64, actionN, true),
    ]);
  }

  forward(x: Tensor): Tensor {
    return this.net.forward(x);
  }
}

function softmax1d(logits: number[]): number[] {
  const max = Math.max(...logits);
  const exps = logits.map(v => Math.exp(v - max));
  const sum = exps.reduce((a, b) => a + b, 0);
  return exps.map(v => v / (sum + 1e-10));
}

function sample(probs: number[]): number {
  const r = Math.random();
  let cum = 0;
  for (let i = 0; i < probs.length; i++) {
    cum += probs[i];
    if (r <= cum) return i;
  }
  return probs.length - 1;
}

function discountedReturns(rewards: number[], gamma: number): number[] {
  const returns = new Array(rewards.length);
  let G = 0;
  for (let t = rewards.length - 1; t >= 0; t--) {
    G = rewards[t] + gamma * G;
    returns[t] = G;
  }
  const mean = returns.reduce((a, b) => a + b, 0) / returns.length;
  const std = Math.sqrt(returns.reduce((s, r) => s + (r - mean) ** 2, 0) / returns.length) + 1e-8;
  return returns.map(r => (r - mean) / std);
}

function computeLoss(
  policyNet: PolicyNet,
  states: number[][],
  actions: number[],
  returns: number[],
): number {
  const T = states.length;
  const flatStates = states.flat();
  const stateTensor = new Tensor(flatStates, [T, 4], false);
  const logits = policyNet.forward(stateTensor);
  const probs = logits.softmax(1);
  let loss = 0;
  for (let t = 0; t < T; t++) {
    const p = probs.data[t * 2 + actions[t]];
    loss -= returns[t] * Math.log(Math.max(p, 1e-10));
  }
  return loss / T;
}

async function playEpisode(
  env: CartPoleEnv,
  policyNet: PolicyNet,
  seed?: number,
  mode?: 'train' | 'eval',
): Promise<{ reward: number; steps: number; states?: number[][]; actions?: number[]; rewards?: number[] }> {
  const { observation } = env.reset({ seed });
  const states: number[][] = [];
  const actions: number[] = [];
  const rewards: number[] = [];
  let obs = observation;
  let totalReward = 0;
  let steps = 0;
  const done = false;

  while (!done) {
    const logits = policyNet.forward(new Tensor(obs, [1, 4], false));
    const probs = softmax1d(logits.data);
    const action = mode === 'eval' ? probs.indexOf(Math.max(...probs)) : sample(probs);

    const result = env.step(action);
    obs = result.observation as number[];

    if (mode === 'train') {
      states.push([...obs]);
      actions.push(action);
      rewards.push(result.reward);
    }

    totalReward += result.reward;
    steps++;

    if (result.terminated || result.truncated) break;
  }

  return mode === 'train'
    ? { reward: totalReward, steps, states, actions, rewards }
    : { reward: totalReward, steps };
}

function learn(
  policyNet: PolicyNet,
  optimizer: Adam,
  states: number[][],
  actions: number[],
  rewards: number[],
  gamma: number,
): void {
  const returns = discountedReturns(rewards, gamma);
  const params = policyNet.parameters();
  const eps = 1e-5;

  for (const p of params) {
    const grad = new Array(p.data.length).fill(0);
    const original = [...p.data];

    for (let i = 0; i < p.data.length; i++) {
      p.data[i] = original[i] + eps;
      const lossPlus = computeLoss(policyNet, states, actions, returns);

      p.data[i] = original[i] - eps;
      const lossMinus = computeLoss(policyNet, states, actions, returns);

      grad[i] = (lossPlus - lossMinus) / (2 * eps);
      p.data[i] = original[i];
    }

    p.grad = grad;
  }

  optimizer.step();
  optimizer.zeroGrad();
}

async function main(): Promise<void> {
  const env = new CartPoleEnv();
  const policyNet = new PolicyNet(4, 2);
  const optimizer = new Adam(policyNet.parameters(), 0.005);
  const gamma = 0.99;
  const maxEpisodes = 1000;

  console.log('='.repeat(55));
  console.log('  CartPole-v1  ·  VPG (REINFORCE)  ·  ai4 nn framework');
  console.log('='.repeat(55));

  // Training
  console.log('\n=== Training ===');
  const episodeRewards: number[] = [];

  for (let ep = 0; ep < maxEpisodes; ep++) {
    const { reward, steps, states, actions, rewards } = await playEpisode(env, policyNet, ep * 100, 'train');

    if (states && actions && rewards) {
      learn(policyNet, optimizer, states, actions, rewards, gamma);
    }

    episodeRewards.push(reward);

    if (ep % 50 === 0 || ep === maxEpisodes - 1) {
      console.log(`  episode ${ep}: reward = ${reward.toFixed(2)}, steps = ${steps}`);
    }

    if (episodeRewards.length >= 20) {
      const avg = episodeRewards.slice(-20).reduce((a, b) => a + b, 0) / 20;
      if (avg > 199) {
        console.log(`\n  Solved at episode ${ep}! Average reward (last 20): ${avg.toFixed(2)}`);
        break;
      }
    }
  }

  // Test
  console.log('\n=== Testing (20 episodes) ===');
  const testRewards: number[] = [];
  for (let i = 0; i < 20; i++) {
    const { reward } = await playEpisode(env, policyNet, i * 1000 + 999, 'eval');
    testRewards.push(reward);
    console.log(`  test ${i + 1}: reward = ${reward.toFixed(2)}`);
  }

  const mean = testRewards.reduce((a, b) => a + b, 0) / testRewards.length;
  const std = Math.sqrt(testRewards.reduce((s, r) => s + (r - mean) ** 2, 0) / testRewards.length);
  console.log(`\n  average test reward = ${mean.toFixed(2)} ± ${std.toFixed(2)}`);
  console.log('='.repeat(55));

  env.close();
}

main().catch(console.error);
